lib/typed_ecto_schema/syntax_sugar.ex

```diff
@@ -116,32 +116,51 @@ defmodule TypedEctoSchema.SyntaxSugar do
     end
   end
 
-  defp transform_expression({:timestamps, ctx, []}) do
-    transform_expression({:timestamps, ctx, [[]]})
+  defp transform_expression({:timestamps, ctx, []}, env) do
+    transform_expression({:timestamps, ctx, [[]]}, env)
   end
 
-  defp transform_expression({:::, _, [{function_name, _, [name, ecto_type, opts]}, type]})
+  defp transform_expression({:"::", _, [{function_name, _, [name, ecto_type, opts]}, type]}, env)
        when function_name in @schema_function_names do
     transform_expression(
-      {function_name, [], [name, ecto_type, [{:__typed_ecto_type__, Macro.escape(type)} | opts]]}
+      {function_name, [], [name, ecto_type, [{:__typed_ecto_type__, Macro.escape(type)} | opts]]},
+      env
     )
   end
 
-  defp transform_expression({:::, _, [{function_name, _, [name, ecto_type]}, type]})
+  defp transform_expression({:"::", _, [{function_name, _, [name, ecto_type]}, type]}, env)
        when function_name in @schema_function_names do
     transform_expression(
-      {function_name, [], [name, ecto_type, [__typed_ecto_type__: Macro.escape(type)]]}
+      {function_name, [], [name, ecto_type, [__typed_ecto_type__: Macro.escape(type)]]},
+      env
     )
   end
 
-  defp transform_expression({:::, _, [{:field, _, [name]}, type]}) do
-    transform_expression({:field, [], [name, :string, [__typed_ecto_type__: Macro.escape(type)]]})
+  defp transform_expression({:"::", _, [{:field, _, [name]}, type]}, env) do
+    transform_expression(
+      {:field, [], [name, :string, [__typed_ecto_type__: Macro.escape(type)]]},
+      env
+    )
   end
 
-  defp transform_expression(other), do: other
+  defp transform_expression(unknown, env) do
+    expanded = Macro.expand(unknown, env)
+
+    case expanded do
+      ^unknown ->
+        unknown
+
+      {:__block__, block_context, calls} ->
+        new_calls = Enum.map(calls, &transform_expression(&1, env))
+        {:__block__, block_context, new_calls}
+
+      call ->
+        transform_expression(call, env)
+    end
```

Comment:
Ok, so this is the core of this PR, right?
You expand whatever call it is and:
- If it's form does not change it means it is not a macro, so you just keep it like that (`unknown`)
- If it expands into a block you get all expressions and transform the expressions recursively
- Else it should be just one expression and then you recursively transform that again

Did I got it right?

Also, the test for the block case is the `add_to_fields` macro, right?